Use KeyStore instead of raw keys with CAs

nklaassen opened the pull request:
This PR updates all uses of private keys held in CAs to use the `KeyStore` interface instead of interacting with the private keys directly, so that the keys may be held in an HSM.

Currently, only the `rawKeyStore` implementation is used (which uses raw in-memory keys with no HSM), so this should result in no functional change.

Remaining steps to implement for full HSM support:
1. Each Auth server with a configured HSM should add its own private keys to each CA.
2. Support for CA rotation with multiple auth servers.
3. User configuration of HSM.

nklaassen committed:
replace GenerateSelfSignedCAWithPrivateKey with GenerateSelfSignedCAWithSigner

nklaassen committed:
Add KeyStore interface with rawKeyStore implementation

nklaassen committed:
add hsmKeyStore implementation

nklaassen committed:
use KeyStore instead of raw keys

nklaassen committed:
vendor go-jose.v2/cryptosigner

nklaassen committed:
switch to all 0s in test uuid

nklaassen committed:
hsm keystore should not fall back to raw keys

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-2

nklaassen committed:
Merge branch 'nklaassen/hsm-2' into nklaassen/hsm-3

nklaassen committed:
Merge branch 'nklaassen/hsm-3' into nklaassen/hsm-4

nklaassen committed:
Merge branch 'nklaassen/hsm-4' into nklaassen/hsm-5

tcsc commented on lib/auth/auth.go:
```diff
@@ -290,6 +297,9 @@ type Server struct {
 	// session related streams
 	streamer events.Streamer
+
+	// keyStore is an interface for interacting with private keys
```
I'd expand this comment a bit - whose private keys, what scope, etc.

nklaassen replied:
I've elaborated a bit here

nklaassen committed:
add comment

nklaassen committed:
Update lib/auth/keystore/keystore.go

Co-authored-by: Brian Joerger <bjoerger@gravitational.com>

nklaassen committed:
use CA constructor in test

nklaassen committed:
remove redundant error returns

nklaassen committed:
Merge branch 'nklaassen/hsm-3' into nklaassen/hsm-4

nklaassen committed:
Merge remote-tracking branch 'origin/nklaassen/hsm-4' into nklaassen/hsm-5

nklaassen committed:
check for existing ID before generating keypair

nklaassen committed:
satisfy linter

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-4

r0mant commented on lib/client/keystore_test.go:
```diff
@@ -428,7 +428,7 @@ func newSelfSignedCA(privateKey []byte) (*tlsca.CertAuthority, auth.TrustedCerts
 		return nil, auth.TrustedCerts{}, trace.Wrap(err)
 	}
-	ca, err := tlsca.FromKeys(cert, privateKey)
+	ca, err := tlsca.FromCertAndSigner(cert, rsaKey.(*rsa.PrivateKey))
```
Check type assertion or is this guaranteed not to panic?

r0mant commented on lib/auth/rotate.go:
```diff
@@ -380,7 +379,7 @@ func (a *Server) autoRotate(ca types.CertAuthority) error {
 // processRotationRequest processes rotation request based on the target and
 // current phase and state.
-func processRotationRequest(req rotationReq) (types.CertAuthority, error) {
+func processRotationRequest(req rotationReq, keyStore keystore.KeyStore) (types.CertAuthority, error) {
```
Nit: Would it make sense to just make this method and startNewRotation members of Server so you don't have to pass an additional function argument?

nklaassen replied:
good point, done in latest commit

r0mant commented on lib/auth/auth.go:
```diff
@@ -497,24 +501,8 @@ func (a *Server) GenerateHostCert(hostPublicKey []byte, hostID, nodeName string,
-		return signer, nil
-	}
-	return nil, trace.NotFound("no raw SSH private key found in CA for %q", ca.GetClusterName())
+func (a *Server) GetKeyStore() keystore.KeyStore {
```
Nit: Missing godoc.

nklaassen committed:
address review comments

- use doc.go instead of README
- add missing license header in hsm.go
- use trace.Wrap
- improve error messages
- return errors from findKeyPair
- factor out findUnusedID function

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-4

nklaassen committed:
import api v7

nklaassen committed:
Merge branch 'nklaassen/hsm-4' into nklaassen/hsm-5

nklaassen committed:
update comments

nklaassen committed:
make rotation helper functions into methods

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-5

nklaassen committed:
Merge branch 'master' into nklaassen/hsm-5
